common/src/main/java/com/wynntils/core/net/UrlManager.java

```diff
@@ -203,33 +263,156 @@ private void downloadAndReadRemoteUrls(URI uri) {
         String localFileName = UrlId.DATA_STATIC_URLS.getId();
 
         Download dl = Managers.Net.download(uri, localFileName);
-        dl.handleInputStream(inputStream -> {
-            try {
-                Pair<Integer, Map<UrlId, UrlInfo>> tryMap = readUrls(inputStream);
-                tryUpdateUrlMap(tryMap, true);
-            } catch (IOException e) {
-                WynntilsMod.warn("Problem updating URL list from online source", e);
-            }
-        });
+        dl.handleInputStream(
+                inputStream -> {
+                    try {
+                        UrlList urlList = readUrls(inputStream);
+                        urlListMap.put(UrlListType.REMOTE, urlList);
+
+                        // Merge the lists with the remote list as the primary source
+                        mergeUrlLists();
+                    } catch (IllegalStateException e) {
+                        // IllegalStateExceptions are thrown on critical errors detected by the manager
+                        WynntilsMod.error("Critical error while updating URL list from online source", e);
+                        throw e;
+                    } catch (IOException e) {
+                        WynntilsMod.warn("Problem updating URL list from online source", e);
+
+                        // Merge the lists we could load, even if the remote list failed
+                        mergeUrlLists();
+                    }
+                },
+                throwable -> WynntilsMod.warn("Failed to download URL list from online source", throwable));
     }
 
-    private void readInputStreamForUrl(InputStream tryStream) throws JsonSyntaxException, IOException {
+    private void readInputStreamForUrl(InputStream tryStream, UrlListType listType)
+            throws JsonSyntaxException, IOException {
         try (InputStream inputStream = tryStream) {
-            Pair<Integer, Map<UrlId, UrlInfo>> tryMap = readUrls(inputStream);
-            tryUpdateUrlMap(tryMap, false);
+            UrlList urlList = readUrls(inputStream);
+            urlListMap.put(listType, urlList);
         } catch (MalformedJsonException e) {
             // This is handled by the caller, so just rethrow
             throw new MalformedJsonException(e);
         }
     }
 
-    private void tryUpdateUrlMap(Pair<Integer, Map<UrlId, UrlInfo>> tryMap, boolean forceUpdateOnSameVersion) {
-        // Even if the version is the same, we might want to update the URLs if the urls are downloaded from the
-        // internet since the URLs might have changed. (local cache and bundled URLs are not always updated)
-        if (tryMap.a() > version || (forceUpdateOnSameVersion && tryMap.a() == version)) {
-            urlMap = tryMap.b();
-            version = tryMap.a();
+    private void mergeUrlLists() {
+        // In theory, this method shouldn't be called concurrently ever, but just in case
+        synchronized (urlMapLock) {
```

Comment:
You don't need a separate object for that! You can just lock on the urlListMap object itself.

Reply:
(That's actually a design decision from Java 1.0 that costs us all a lot these days -- all objects need to reserve bits in the object header since it could potentially be used as a lock...)